## Project  Title - IT Project Delivery Performance Analysis of Tech Mahindra

In [2]:
import pandas as pd 
import numpy as np 

In [3]:
df = pd.read_csv("TechMahindra_Project_Delivery_Data_100_Rows.csv")

In [4]:
df.head()

Project_ID              Client_Name  Start_Date Planned_End_Date  \
0     PRJ001               Brooks Ltd  2024-05-11       2025-04-23   
1     PRJ002  Weiss, Jordan and Jones  2025-01-09       2025-02-04   
2     PRJ003            Beard-Hoffman  2024-11-01       2025-01-06   
3     PRJ004           Cooper-Sanchez  2024-11-30       2025-05-24   
4     PRJ005                Smith PLC  2024-10-16       2024-10-17   

     End_Date  Project_Budget  Actual_Cost  Team_Size       Status  \
0  2025-05-09          389015       414164         13      Delayed   
1  2025-03-22          361134       434837         11    Completed   
2  2025-01-06          188999       229205         10    Completed   
3  2025-06-23          192053       154046          6  In Progress   
4  2025-03-31          479752       600905         10    Completed   

  Technology_Stack  
0       Salesforce  
1             .NET  
2             .NET  
3             .NET  
4             Java

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Project_ID        100 non-null    object
 1   Client_Name       100 non-null    object
 2   Start_Date        100 non-null    object
 3   Planned_End_Date  100 non-null    object
 4   End_Date          100 non-null    object
 5   Project_Budget    100 non-null    int64 
 6   Actual_Cost       100 non-null    int64 
 7   Team_Size         100 non-null    int64 
 8   Status            100 non-null    object
 9   Technology_Stack  100 non-null    object
dtypes: int64(3), object(7)
memory usage: 7.9+ KB


In [6]:
df.any()

Project_ID          True
Client_Name         True
Start_Date          True
Planned_End_Date    True
End_Date            True
Project_Budget      True
Actual_Cost         True
Team_Size           True
Status              True
Technology_Stack    True
dtype: bool

In [7]:
df.isnull().sum()

Project_ID          0
Client_Name         0
Start_Date          0
Planned_End_Date    0
End_Date            0
Project_Budget      0
Actual_Cost         0
Team_Size           0
Status              0
Technology_Stack    0
dtype: int64

In [8]:
df.describe()

Project_Budget    Actual_Cost   Team_Size
count      100.000000     100.000000  100.000000
mean    317064.090000  345187.080000   12.080000
std     119830.364788  148359.593254    5.012469
min     103835.000000   85347.000000    4.000000
25%     210060.250000  228334.250000    8.000000
50%     330827.500000  334739.500000   12.000000
75%     431017.500000  458865.750000   17.000000
max     494278.000000  619151.000000   20.000000

In [9]:
df.head()

Project_ID              Client_Name  Start_Date Planned_End_Date  \
0     PRJ001               Brooks Ltd  2024-05-11       2025-04-23   
1     PRJ002  Weiss, Jordan and Jones  2025-01-09       2025-02-04   
2     PRJ003            Beard-Hoffman  2024-11-01       2025-01-06   
3     PRJ004           Cooper-Sanchez  2024-11-30       2025-05-24   
4     PRJ005                Smith PLC  2024-10-16       2024-10-17   

     End_Date  Project_Budget  Actual_Cost  Team_Size       Status  \
0  2025-05-09          389015       414164         13      Delayed   
1  2025-03-22          361134       434837         11    Completed   
2  2025-01-06          188999       229205         10    Completed   
3  2025-06-23          192053       154046          6  In Progress   
4  2025-03-31          479752       600905         10    Completed   

  Technology_Stack  
0       Salesforce  
1             .NET  
2             .NET  
3             .NET  
4             Java

In [10]:
df["Start_Date"]=pd.to_datetime(df["Start_Date"])
df["Planned_End_Date"]=pd.to_datetime(df["Planned_End_Date"])
df["End_Date"]=pd.to_datetime(df["End_Date"])

In [11]:
df.head()

Project_ID              Client_Name Start_Date Planned_End_Date   End_Date  \
0     PRJ001               Brooks Ltd 2024-05-11       2025-04-23 2025-05-09   
1     PRJ002  Weiss, Jordan and Jones 2025-01-09       2025-02-04 2025-03-22   
2     PRJ003            Beard-Hoffman 2024-11-01       2025-01-06 2025-01-06   
3     PRJ004           Cooper-Sanchez 2024-11-30       2025-05-24 2025-06-23   
4     PRJ005                Smith PLC 2024-10-16       2024-10-17 2025-03-31   

   Project_Budget  Actual_Cost  Team_Size       Status Technology_Stack  
0          389015       414164         13      Delayed       Salesforce  
1          361134       434837         11    Completed             .NET  
2          188999       229205         10    Completed             .NET  
3          192053       154046          6  In Progress             .NET  
4          479752       600905         10    Completed             Java

In [12]:
df['Delivery_Delay'] =(df['End_Date']-df['Planned_End_Date']).dt.days

In [13]:
df.head()

Project_ID              Client_Name Start_Date Planned_End_Date   End_Date  \
0     PRJ001               Brooks Ltd 2024-05-11       2025-04-23 2025-05-09   
1     PRJ002  Weiss, Jordan and Jones 2025-01-09       2025-02-04 2025-03-22   
2     PRJ003            Beard-Hoffman 2024-11-01       2025-01-06 2025-01-06   
3     PRJ004           Cooper-Sanchez 2024-11-30       2025-05-24 2025-06-23   
4     PRJ005                Smith PLC 2024-10-16       2024-10-17 2025-03-31   

   Project_Budget  Actual_Cost  Team_Size       Status Technology_Stack  \
0          389015       414164         13      Delayed       Salesforce   
1          361134       434837         11    Completed             .NET   
2          188999       229205         10    Completed             .NET   
3          192053       154046          6  In Progress             .NET   
4          479752       600905         10    Completed             Java   

   Delivery_Delay  
0              16  
1              46  
2               0  
3              30  
4             165

In [14]:
df['Cost_Overrun'] =(df['Actual_Cost']-df['Project_Budget'])

In [15]:
df.head()

Project_ID              Client_Name Start_Date Planned_End_Date   End_Date  \
0     PRJ001               Brooks Ltd 2024-05-11       2025-04-23 2025-05-09   
1     PRJ002  Weiss, Jordan and Jones 2025-01-09       2025-02-04 2025-03-22   
2     PRJ003            Beard-Hoffman 2024-11-01       2025-01-06 2025-01-06   
3     PRJ004           Cooper-Sanchez 2024-11-30       2025-05-24 2025-06-23   
4     PRJ005                Smith PLC 2024-10-16       2024-10-17 2025-03-31   

   Project_Budget  Actual_Cost  Team_Size       Status Technology_Stack  \
0          389015       414164         13      Delayed       Salesforce   
1          361134       434837         11    Completed             .NET   
2          188999       229205         10    Completed             .NET   
3          192053       154046          6  In Progress             .NET   
4          479752       600905         10    Completed             Java   

   Delivery_Delay  Cost_Overrun  
0              16         25149  
1              46         73703  
2               0         40206  
3              30        -38007  
4             165        121153

In [16]:
df.value_counts(['Status'])

Status     
Delayed        39
Completed      36
In Progress    25
Name: count, dtype: int64

In [17]:
Total_Project = df['Project_ID'].nunique()

In [18]:
Total_Overrun_Cost =df["Cost_Overrun"].sum()

In [19]:
# Average Cost Overrun Per Project 
Average_Cost_Overrun_Per_project = Total_Overrun_Cost / Total_Project

In [20]:
print("Average Cost Overrun Per project is : ",Average_Cost_Overrun_Per_project)

Average Cost Overrun Per project is :  28122.99


EDA for few Questions 

# Q1. What % of Projects were delayed 

In [21]:

Total_Project

100

In [22]:
df['Status'].value_counts()

Status
Delayed        39
Completed      36
In Progress    25
Name: count, dtype: int64

In [23]:
delayed_count =(df['Status']=='Delayed').sum()

In [24]:
delayed_count

39

In [25]:
Percent_Proj_Delay =(delayed_count/Total_Project)*100

In [26]:
print("Total percentage of Project were Delayed :",Percent_Proj_Delay,"%")

Total percentage of Project were Delayed : 39.0 %


# Ans. - 39% of Projects were Delayed 

# Q2. Which technology stack has the most delays?


In [27]:
df.head()

Project_ID              Client_Name Start_Date Planned_End_Date   End_Date  \
0     PRJ001               Brooks Ltd 2024-05-11       2025-04-23 2025-05-09   
1     PRJ002  Weiss, Jordan and Jones 2025-01-09       2025-02-04 2025-03-22   
2     PRJ003            Beard-Hoffman 2024-11-01       2025-01-06 2025-01-06   
3     PRJ004           Cooper-Sanchez 2024-11-30       2025-05-24 2025-06-23   
4     PRJ005                Smith PLC 2024-10-16       2024-10-17 2025-03-31   

   Project_Budget  Actual_Cost  Team_Size       Status Technology_Stack  \
0          389015       414164         13      Delayed       Salesforce   
1          361134       434837         11    Completed             .NET   
2          188999       229205         10    Completed             .NET   
3          192053       154046          6  In Progress             .NET   
4          479752       600905         10    Completed             Java   

   Delivery_Delay  Cost_Overrun  
0              16         25149  
1              46         73703  
2               0         40206  
3              30        -38007  
4             165        121153

In [28]:
df.value_counts(['Technology_Stack'])

Technology_Stack
Java                19
React               15
Salesforce          13
Node.js             12
Python              12
Angular             11
.NET                10
SAP                  8
Name: count, dtype: int64

In [29]:
# Which technology stack has the most delays?
delayed_proj = df[df['Status']=='Delayed']

In [30]:
delayed_proj

Project_ID                    Client_Name Start_Date Planned_End_Date  \
0      PRJ001                     Brooks Ltd 2024-05-11       2025-04-23   
7      PRJ008         Smith, Moore and Davis 2023-09-22       2023-11-08   
9      PRJ010                 Bautista Group 2025-04-01       2025-05-04   
10     PRJ011                     Ross-Davis 2024-07-30       2024-09-05   
14     PRJ015                     Estes-York 2023-12-16       2025-01-16   
19     PRJ020                     Knight Inc 2024-02-12       2025-03-12   
21     PRJ022                     Maddox PLC 2024-02-05       2025-02-24   
27     PRJ028                     Santos Ltd 2025-06-24       2025-06-26   
29     PRJ030      Nunez, Carlson and Wilson 2025-06-22       2025-07-02   
32     PRJ033        White, Graham and White 2024-05-22       2025-02-02   
35     PRJ036     Moore, Henderson and Blair 2023-08-13       2024-02-22   
38     PRJ039               Williams-Stewart 2025-06-02       2025-06-05   
49     PRJ050      Hudson, Miller and Butler 2024-10-06       2025-01-12   
52     PRJ053                   Waters-Clark 2024-09-07       2025-02-26   
54     PRJ055                    Roberts PLC 2025-05-06       2025-06-15   
55     PRJ056                Alvarez-Gilmore 2024-04-05       2025-01-24   
57     PRJ058   Zamora, Daniels and Phillips 2023-07-19       2024-12-21   
58     PRJ059               Murillo and Sons 2024-08-07       2024-11-10   
59     PRJ060    Carter, Mendoza and Brennan 2024-05-14       2024-10-01   
60     PRJ061                      Burke PLC 2025-02-19       2025-03-10   
62     PRJ063      Davis, Fernandez and Ryan 2025-05-16       2025-05-30   
63     PRJ064      Gregory, Poole and Rogers 2023-11-04       2024-12-02   
64     PRJ065  Morales, Johnson and Villegas 2024-04-22       2024-09-17   
65     PRJ066                     Brown-Hill 2024-02-20       2025-06-30   
68     PRJ069         Smith, Graham and Howe 2024-05-13       2024-09-19   
72     PRJ073      Ward, Johnson and Ramirez 2024-09-15       2025-05-05   
73     PRJ074     Mcintyre, Wilkins and Kent 2024-09-12       2025-06-12   
74     PRJ075                      Lynch Inc 2025-03-27       2025-04-05   
75     PRJ076                 White and Sons 2024-04-06       2024-06-08   
77     PRJ078               Bullock and Sons 2025-03-29       2025-07-01   
81     PRJ082                   Clark-Brewer 2024-12-09       2024-12-18   
82     PRJ083    Ferguson, Holland and Davis 2023-11-17       2024-10-23   
86     PRJ087    Williams, Warren and Booker 2024-09-02       2025-01-11   
88     PRJ089                     Holmes Ltd 2025-02-25       2025-06-07   
89     PRJ090                  Sanchez-Jones 2024-11-05       2024-11-10   
93     PRJ094                  Vargas-Forbes 2023-07-28       2024-06-27   
94     PRJ095                Campos and Sons 2024-05-02       2024-12-04   
97     PRJ098        Price, Martin and Quinn 2024-12-30       2025-04-26   
98     PRJ099              Robinson and Sons 2024-02-27       2025-05-01   

     End_Date  Project_Budget  Actual_Cost  Team_Size   Status  \
0  2025-05-09          389015       414164         13  Delayed   
7  2023-11-08          335089       363910         18  Delayed   
9  2025-05-04          324054       420483          4  Delayed   
10 2024-09-05          440352       548805          9  Delayed   
14 2025-01-16          446065       464751          5  Delayed   
19 2025-03-12          301472       256699          7  Delayed   
21 2025-02-24          118413       128477          8  Delayed   
27 2025-06-26          122198        99398          9  Delayed   
29 2025-07-03          405775       368219          9  Delayed   
32 2025-02-02          278879       272803         10  Delayed   
35 2024-08-02          177649       174096          5  Delayed   
38 2025-06-05          292424       314440         11  Delayed   
49 2025-04-13          151089       129710         10  Delayed   
52 2025-04-19          453333       58

In [31]:
Technology_Delayed = delayed_proj['Technology_Stack'].value_counts()

In [32]:
Technology_Delayed

Technology_Stack
Java          9
Salesforce    8
Angular       6
React         5
Node.js       4
Python        3
SAP           2
.NET          2
Name: count, dtype: int64

In [33]:
delay_count =Technology_Delayed.max()
delay_count

9

In [34]:
most_delayed_tech = Technology_Delayed.idxmax()
most_delayed_tech


'Java'

In [35]:
print(f"Technology with Most Delays : {most_delayed_tech} ({delay_count} delays)")

Technology with Most Delays : Java (9 delays)


# Ans. Java Technology have most Delays (9)

# Q3. Which clients have the highest cost overruns?


In [36]:
df.head()

Project_ID              Client_Name Start_Date Planned_End_Date   End_Date  \
0     PRJ001               Brooks Ltd 2024-05-11       2025-04-23 2025-05-09   
1     PRJ002  Weiss, Jordan and Jones 2025-01-09       2025-02-04 2025-03-22   
2     PRJ003            Beard-Hoffman 2024-11-01       2025-01-06 2025-01-06   
3     PRJ004           Cooper-Sanchez 2024-11-30       2025-05-24 2025-06-23   
4     PRJ005                Smith PLC 2024-10-16       2024-10-17 2025-03-31   

   Project_Budget  Actual_Cost  Team_Size       Status Technology_Stack  \
0          389015       414164         13      Delayed       Salesforce   
1          361134       434837         11    Completed             .NET   
2          188999       229205         10    Completed             .NET   
3          192053       154046          6  In Progress             .NET   
4          479752       600905         10    Completed             Java   

   Delivery_Delay  Cost_Overrun  
0              16         25149  
1              46         73703  
2               0         40206  
3              30        -38007  
4             165        121153

Which clients have the highest cost overruns?


In [37]:
Highest_Costoverrun =df['Cost_Overrun'].max()
Highest_Costoverrun

141316

In [38]:
Client_Cost_Overrun = df[df['Cost_Overrun']==Highest_Costoverrun]

In [39]:
Client_Cost_Overrun

Project_ID Client_Name Start_Date Planned_End_Date   End_Date  \
91     PRJ092  Love Group 2024-12-16       2025-03-20 2025-03-20   

    Project_Budget  Actual_Cost  Team_Size       Status Technology_Stack  \
91          477835       619151          8  In Progress           Python   

    Delivery_Delay  Cost_Overrun  
91               0        141316

In [40]:

print(Client_Cost_Overrun[['Client_Name','Cost_Overrun']])


   Client_Name  Cost_Overrun
91  Love Group        141316


In [41]:
Client_Name =Client_Cost_Overrun['Client_Name'].values[0]

In [42]:
print("Client which have the most Cost Overrun is : ",Client_Name)

Client which have the most Cost Overrun is :  Love Group


# Ans. Client Love Group have the most Costoverrun 

# Q4. How does team size impact delivery?


In [43]:
df.head()

Project_ID              Client_Name Start_Date Planned_End_Date   End_Date  \
0     PRJ001               Brooks Ltd 2024-05-11       2025-04-23 2025-05-09   
1     PRJ002  Weiss, Jordan and Jones 2025-01-09       2025-02-04 2025-03-22   
2     PRJ003            Beard-Hoffman 2024-11-01       2025-01-06 2025-01-06   
3     PRJ004           Cooper-Sanchez 2024-11-30       2025-05-24 2025-06-23   
4     PRJ005                Smith PLC 2024-10-16       2024-10-17 2025-03-31   

   Project_Budget  Actual_Cost  Team_Size       Status Technology_Stack  \
0          389015       414164         13      Delayed       Salesforce   
1          361134       434837         11    Completed             .NET   
2          188999       229205         10    Completed             .NET   
3          192053       154046          6  In Progress             .NET   
4          479752       600905         10    Completed             Java   

   Delivery_Delay  Cost_Overrun  
0              16         25149  
1              46         73703  
2               0         40206  
3              30        -38007  
4             165        121153

In [44]:
delayed = df[df['Status']=='Delayed']['Team_Size']
on_time = df[df['Status']=='Completed']['Team_Size']

In [45]:
print("Average team size(Delayed) :",delayed.mean())
print("Average team size (on_time) :",on_time.mean())

Average team size(Delayed) : 11.846153846153847
Average team size (on_time) : 12.416666666666666


# Larger teams may contribute slightly to better on-time delivery, but the impact is not very strong in this dataset.

In [46]:
bins = [0,8,12, float('inf')]
labels =['Small', 'Medium','Large']

df['Team_Size_Category'] = pd.cut(df['Team_Size'], bins = bins , labels=labels)

print(df[['Team_Size','Team_Size_Category']].head())

   Team_Size Team_Size_Category
0         13              Large
1         11             Medium
2         10             Medium
3          6              Small
4         10             Medium


In [47]:
df.head(2)

Project_ID              Client_Name Start_Date Planned_End_Date   End_Date  \
0     PRJ001               Brooks Ltd 2024-05-11       2025-04-23 2025-05-09   
1     PRJ002  Weiss, Jordan and Jones 2025-01-09       2025-02-04 2025-03-22   

   Project_Budget  Actual_Cost  Team_Size     Status Technology_Stack  \
0          389015       414164         13    Delayed       Salesforce   
1          361134       434837         11  Completed             .NET   

   Delivery_Delay  Cost_Overrun Team_Size_Category  
0              16         25149              Large  
1              46         73703             Medium

In [48]:
Avg_Delay_by_team_size =df.groupby('Team_Size_Category')['Delivery_Delay'].mean()
print("Avrage Delivery Delay by Team Size Category :\n",Avg_Delay_by_team_size)

Avrage Delivery Delay by Team Size Category :
 Team_Size_Category
Small     13.103448
Medium    46.692308
Large     19.600000
Name: Delivery_Delay, dtype: float64


C:\Users\Naresh\AppData\Local\Temp\ipykernel_16852\3279776480.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  Avg_Delay_by_team_size =df.groupby('Team_Size_Category')['Delivery_Delay'].mean()


In [49]:
correlation = df[['Team_Size','Delivery_Delay']].corr()

In [50]:
print("Correlation : \n" , correlation)

Correlation : 
                 Team_Size  Delivery_Delay
Team_Size        1.000000       -0.004631
Delivery_Delay  -0.004631        1.000000


# Ans. Projects with larger teams  have lower average delivery delays.Average delay decreased from Small → Medium → Large teams, and a negative correlation of -0.004631 between team size and delivery delay confirms that larger teams are associated with faster delivery.

In [51]:
print("Insights From the Data")

Insights From the Data


In [55]:
# Client With Zero Cost Overrun 
Zero_cost_clients =df[df['Cost_Overrun']==0]['Client_Name'].unique()
print(f"{Zero_cost_clients} clients had zero or nagative cost overrun")

[] clients had zero or nagative cost overrun


In [56]:
df[['Delivery_Delay','Cost_Overrun']].corr()

Delivery_Delay  Cost_Overrun
Delivery_Delay        1.000000      0.157382
Cost_Overrun          0.157382      1.000000

# Cost Overrun by Technology Stack

In [57]:
df.groupby('Technology_Stack')['Cost_Overrun'].mean().sort_values(ascending=False)

Technology_Stack
Angular       37431.363636
Salesforce    32091.076923
Python        31336.416667
SAP           29392.000000
React         28945.933333
Java          24987.473684
Node.js       24204.583333
.NET          17279.100000
Name: Cost_Overrun, dtype: float64

# Project Delayed Beyond 60 Days

In [59]:
severely_delayed =df[df['Delivery_Delay']>60]
Percent =(len(severely_delayed)/len(df))*100
print(f"{Percent:.2f}% of projects delayed by more than 60 days")

13.00% of projects delayed by more than 60 days


# Average Team Size by Project Status

In [60]:
df.groupby('Status')['Team_Size'].mean()

Status
Completed      12.416667
Delayed        11.846154
In Progress    11.960000
Name: Team_Size, dtype: float64

# Insights From the Dataset 

## Q1. What % of Projects were delayed 
### Ans. 39% of Projects were Delayed

## Q2. Which technology stack has the most delays?
### Ans. Java Technology have most Delays (9)

## Q3. Which clients have the highest cost overruns?
### Ans. Client which have the most Cost Overrun is :  Love Group

## Q4. How does team size impact delivery?
### Ans.Projects with larger teams  have lower average delivery delays.Average delay decreased from Small → Medium → Large teams, and a negative correlation of -0.004631 between team size and delivery delay confirms that larger teams are associated with faster delivery.